<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Bioconductor toolkit for single-cell RNA velocity </title>
    <meta charset="utf-8" />
    <meta name="author" content="Kevin Rue-Albrecht, Charlotte Soneson, Michael Stadler, Aaron Lun" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css/rladies-fonts.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome/css/v4-shims.css" rel="stylesheet" />
    <link rel="stylesheet" href="my-theme.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Bioconductor toolkit for single-cell RNA velocity <img src="img/velociraptor_sticker.png" height="200px" style="vertical-align:bottom;float:right">
## Bioconductor Conference 2021
### <a href="https://kevinrue.github.io">Kevin Rue-Albrecht</a>, <a href="https://csoneson.github.io">Charlotte Soneson</a>, <a href="https://github.com/mbstadler">Michael Stadler</a>, <a href="https://github.com/LTLA">Aaron Lun</a>
### 2021-08-06 (updated: 2021-07-21)

---


layout: true

&lt;div class="my-header"&gt;&lt;img src="img/velociraptor_sticker.png" alt="logo" align="right" height="90%"&gt;&lt;/div&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;
Kevin Rue-Albrecht
&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;&amp;emsp;
velociraptor
&lt;/span&gt;&lt;/div&gt;











---

# RNA velocity predicts the future state of cells

&lt;img src="img/lamanno2018_fig3.png" width="2849" /&gt;

.right[
.small-p[
RNA velocity field describes fate decisions of major neural lineages in the hippocampus.
&lt;a name=cite-lamanno2018&gt;&lt;/a&gt;([La Manno, et al., 2018](https://doi.org/10.1038/s41586-018-0414-6))
]
]

---

# RNA velocity from spliced and unspliced mRNAs

&lt;img src="img/lamanno2018_fig1.png" width="2140" /&gt;

.right[
.small-p[
Balance between unspliced and spliced mRNAs is predictive of cellular state progression.
([La Manno, et al., 2018](https://doi.org/10.1038/s41586-018-0414-6))
]
]

---

# scVelo

* Python package
  + [Read the Docs](https://scvelo.readthedocs.io/).

* Estimates RNA velocity from estimates of gene-wise spliced and unspliced abundances.

* Steady-state (deterministic), stochastic, dynamic models.
  + [Read the Docs](https://scvelo.readthedocs.io/about/#rna-velocity-models)

* Uses the &lt;i class="fab fa-python"&gt;&lt;/i&gt; `AnnData` class for internal data representation
  + [Read the Docs](https://anndata.readthedocs.io/en/latest/)

* Not straightforward to integrate in an R/Bioconductor-based workflow.

.right[
.small-p[
Generalizing RNA velocity to transient cell states through dynamical modeling
&lt;a name=cite-bergen2020&gt;&lt;/a&gt;([Bergen, et al., 2020](https://www.ncbi.nlm.nih.gov/pubmed/32747759))
]
]

---

# The Bioconductor factor

![](img/distracted-analyst.jpg)&lt;!-- --&gt;

---

# &lt;img src="img/velociraptor_sticker.png" height="70px" style="vertical-align:bottom"&gt; is a Bioconductor-friendly wrapper of &lt;i class="fab fa-python"&gt;&lt;/i&gt; scVelo

&lt;img src="img/bioconductor_sticker.png" height="30px" style="vertical-align:bottom"&gt; *[velociraptor](https://bioconductor.org/packages/3.14/velociraptor)*
uses
&lt;img src="img/bioconductor_sticker.png" height="30px" style="vertical-align:bottom"&gt; *[basilisk](https://bioconductor.org/packages/3.14/basilisk)*
to run &lt;i class="fab fa-python"&gt;&lt;/i&gt; [scVelo](https://pypi.org/project/scvelo/)
in a `BasiliskEnvironment`
(i.e., a &lt;img src="img/conda.png" height="30px" style="vertical-align:bottom"&gt; [Conda](https://docs.conda.io/en/latest/) environment).


```r
setMethod("scvelo", "SummarizedExperiment", function(x, ...,
    assay.X="counts", assay.spliced="spliced", assay.unspliced="unspliced")
{
  ...
  output &lt;- basiliskRun(env = velo.env, fun = .run_scvelo,
      X = X, spliced = spliced, unspliced = unspliced,
      use.theirs = use.theirs, mode = mode,
      scvelo.params = scvelo.params,
      dimred = dimred)
  ...
}
```

`velociraptor::scvelo()` returns a `SingleCellExperiment` object containing the output of the velocity calculations.

---

# Input and data representation

* The input to &lt;i class="fab fa-python"&gt;&lt;/i&gt; [scVelo](https://pypi.org/project/scvelo/) is two gene-by-cell count matrices, tabulating the UMI count for spliced and unspliced variants of each gene.
* Several tools exist to estimate these counts from the raw reads &lt;a name=cite-soneson2021&gt;&lt;/a&gt;([Soneson, et al., 2021](https://www.ncbi.nlm.nih.gov/pubmed/33428615)).
* Count matrices are stored as assays in a `SummarizedExperiment` object.
* The &lt;img src="img/bioconductor_sticker.png" height="30px" style="vertical-align:bottom"&gt; *[zellkonverter](https://bioconductor.org/packages/3.14/zellkonverter)* package takes care of conversion between the `AnnData` and `SingleCellExperiment` formats.



.small-code[

```r
hermann
```

```
## class: SingleCellExperiment 
## dim: 54448 1711 
## metadata(0):
## assays(3): spliced unspliced logcounts
## rownames(54448): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...
##   ENSMUSG00000064369.1 ENSMUSG00000064372.1
## rowData names(0):
## colnames(1711): TGACAACAGGACAGAA TTGGAACAGGCGTACA ... TCGCGTTCAAGAGTCG
##   CACCTTGCAGATCGGA
## colData names(2): celltype sizeFactor
## reducedDimNames(2): PCA TSNE
## mainExpName: NULL
## altExpNames(0):
```
]

---

# Running velociraptor::scvelo()

Use gene selection, normalization and dimension reduction from &lt;i class="fab fa-python"&gt;&lt;/i&gt; [scVelo](https://pypi.org/project/scvelo/), or customize each step in &lt;i class="fab fa-r-project"&gt;&lt;/i&gt; for full compatibility with the rest of the workflow.

.small-code[

```r
hermann_velo &lt;- velociraptor::scvelo(hermann, subset.row = top.hvgs, assay.X = "spliced", use.dimred = "PCA")
hermann_velo
```

```
## class: SingleCellExperiment 
## dim: 2000 1711 
## metadata(4): neighbors velocity_params velocity_graph
##   velocity_graph_neg
## assays(6): X spliced ... Mu velocity
## rownames(2000): ENSMUSG00000038015.6 ENSMUSG00000022501.6 ...
##   ENSMUSG00000115935.1 ENSMUSG00000029725.10
## rowData names(3): velocity_gamma velocity_r2 velocity_genes
## colnames(1711): TGACAACAGGACAGAA TTGGAACAGGCGTACA ... TCGCGTTCAAGAGTCG
##   CACCTTGCAGATCGGA
## colData names(7): velocity_self_transition root_cells ...
##   velocity_confidence velocity_confidence_transition
## reducedDimNames(1): X_pca
## mainExpName: NULL
## altExpNames(0):
```

```r
reducedDim(hermann_velo, "TSNE") &lt;- reducedDim(hermann, "TSNE")
hermann_velo$celltype &lt;- hermann$celltype
```
]

---

# Project velocities onto low-dimensional embedding



```r
embedded &lt;- embedVelocity(reducedDim(hermann, "TSNE"), hermann_velo)
grid.df &lt;- gridVectors(reducedDim(hermann, "TSNE"), embedded)

plotTSNE(hermann, colour_by="celltype") +
    geom_segment(data = grid.df, mapping = aes(x = start.1, y = start.2, 
        xend = end.1, yend = end.2), arrow = arrow(length = unit(0.05, "inches")))
```

&lt;img src="index_files/figure-html/unnamed-chunk-9-1.png" style="display: block; margin: auto;" /&gt;

---

# Gene-wise visualizations


```r
plotVelocity(hermann_velo, c("ENSMUSG00000032601.13"), use.dimred = "TSNE", 
             color_by = "celltype")
```

&lt;img src="index_files/figure-html/unnamed-chunk-10-1.png" style="display: block; margin: auto;" /&gt;

---

# Stream lines


```r
plotVelocityStream(hermann_velo, embedded, use.dimred = "TSNE", 
  color.streamlines = TRUE, color_by = "velocity_pseudotime")
```

&lt;img src="index_files/figure-html/unnamed-chunk-11-1.png" style="display: block; margin: auto;" /&gt;


---

# Processing a dataset for RNA velocity analysis

* The Hermann spermatogenesis data set used in these slides is available from the &lt;img src="img/bioconductor_sticker.png" height="30px" style="vertical-align:bottom"&gt; *[scRNAseq](https://bioconductor.org/packages/3.14/scRNAseq)* package.
* The script detailing the pre-processing is available [here](https://github.com/LTLA/scRNAseq/blob/master/inst/scripts/2.4.0/make-hermann-spermatogenesis-data.Rmd) &lt;sup&gt;1&lt;/sup&gt;
* The source code for this presentation is available [here](https://github.com/kevinrue/velociraptor_talk_bioc2021) &lt;sup&gt;2&lt;/sup&gt;

.center[
&lt;img src="img/scrnaseq-github.png" height="300px"&gt;
]

.small-p[
&lt;sup&gt;1&lt;/sup&gt; Long URL: &lt;https://github.com/LTLA/scRNAseq/blob/master/inst/scripts/2.4.0/make-hermann-spermatogenesis-data.Rmd&gt;

&lt;sup&gt;2&lt;/sup&gt; Long URL: &lt;https://github.com/kevinrue/velociraptor_talk_bioc2021&gt;
]

---

# References

.small-p[
&lt;a name=bib-bergen2020&gt;&lt;/a&gt;[Bergen, V. et al.](#cite-bergen2020)
(2020). "Generalizing RNA velocity to transient cell states through
dynamical modeling". In: _Nat Biotechnol_ 38.12, pp. 1408-1414. ISSN:
1546-1696 (Electronic) 1087-0156 (Linking). DOI:
[10.1038/s41587-020-0591-3](https://doi.org/10.1038%2Fs41587-020-0591-3).
URL:
[https://www.ncbi.nlm.nih.gov/pubmed/32747759](https://www.ncbi.nlm.nih.gov/pubmed/32747759).

&lt;a name=bib-orchestrating2015&gt;&lt;/a&gt;[Huber, W. et
al.](#cite-orchestrating2015) (2015). "Orchestrating high-throughput
genomic analysis with Bioconductor". In: _Nat Methods_ 12.2, pp.
115-21. ISSN: 1548-7105 (Electronic) 1548-7091 (Linking). DOI:
[10.1038/nmeth.3252](https://doi.org/10.1038%2Fnmeth.3252). URL:
[https://www.ncbi.nlm.nih.gov/pubmed/25633503](https://www.ncbi.nlm.nih.gov/pubmed/25633503).

&lt;a name=bib-lamanno2018&gt;&lt;/a&gt;[La Manno, G. et al.](#cite-lamanno2018)
(2018). "RNA velocity of single cells". In: _Nature_ 560.7719, pp.
494-498. DOI:
[10.1038/s41586-018-0414-6](https://doi.org/10.1038%2Fs41586-018-0414-6).
URL:
[https://doi.org/10.1038/s41586-018-0414-6](https://doi.org/10.1038/s41586-018-0414-6).

&lt;a name=bib-soneson2021&gt;&lt;/a&gt;[Soneson, C. et al.](#cite-soneson2021)
(2021). "Preprocessing choices affect RNA velocity results for droplet
scRNA-seq data". In: _PLoS Comput Biol_ 17.1, p. e1008585. ISSN:
1553-7358 (Electronic) 1553-734X (Linking). DOI:
[10.1371/journal.pcbi.1008585](https://doi.org/10.1371%2Fjournal.pcbi.1008585).
URL:
[https://www.ncbi.nlm.nih.gov/pubmed/33428615](https://www.ncbi.nlm.nih.gov/pubmed/33428615).
]

.center[
&lt;img src="img/bioconductor_sticker.png" height="300px"&gt;
&lt;img src="img/velociraptor_sticker.png" height="300px"&gt;
]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
